File: test/bench/java/deriv.java
interface Expr {
}

final class ValExpr implements Expr {
  long value;
  ValExpr(long i) {
    value = i;
  }
}

final class VarExpr implements Expr {
  String name;
  VarExpr(String s) {
    name = s;
  }
}

final class LnExpr implements Expr {
  Expr expr;
  LnExpr(Expr e) {
    expr = e;
  }
}

final class AddExpr implements Expr {
  Expr left;
  Expr right;
  AddExpr(Expr l, Expr r) {
    left = l;
    right = r;
  }
}

final class MulExpr implements Expr {
  Expr left;
  Expr right;
  MulExpr(Expr l, Expr r) {
    left = l;
    right = r;
  }
}

final class PowExpr implements Expr {
  Expr left;
  Expr right;
  PowExpr(Expr l, Expr r) {
    left = l;
    right = r;
  }
}



public class deriv {
  static long pown(long x, long n) {
    if (n==0) return 1;
    else if (n == 1) return x;
    else {
      long y = pown(x, n/2);
      return (y * y * (n%2 == 0 ? 1 : x));
    }
  }

  static Expr add( Expr x, Expr y ) {
    if (x instanceof ValExpr a && y instanceof ValExpr b) {
      return new ValExpr( a.value + b.value );
    }
    else if (x instanceof ValExpr a && a.value == 0) {
      return y;
    }
    else if (y instanceof ValExpr b && b.value == 0) {
      return x;
    }
    else if (y instanceof ValExpr b) {
      return add(y,x);
    }
    else if (x instanceof ValExpr a && y instanceof AddExpr b && b.left instanceof ValExpr bl) {
      return add(new ValExpr(a.value + bl.value), b.right);
    }
    else if (y instanceof AddExpr b && b.left instanceof ValExpr) {
      return add(b.left, add(x,b.right));
    }
    else if (x instanceof AddExpr a) {
      return add(a.left, add(a.right,y));
    }
    else {
      return new AddExpr(x,y);
    }
  }

  static Expr mul( Expr x, Expr y ) {
    if (x instanceof ValExpr a && y instanceof ValExpr b) {
      return new ValExpr( a.value * b.value );
    }
    else if (x instanceof ValExpr a && a.value == 0) {
      return x;
    }
    else if (y instanceof ValExpr b && b.value == 0) {
      return y;
    }
    else if (x instanceof ValExpr a && a.value == 1) {
      return y;
    }
    else if (y instanceof ValExpr b && b.value == 1) {
      return x;
    }
    else if (y instanceof ValExpr b) {
      return mul(y,x);
    }
    else if (x instanceof ValExpr a && y instanceof MulExpr b && b.left instanceof ValExpr bl) {
      return mul(new ValExpr(a.value * bl.value), b.right);
    }
    else if (y instanceof MulExpr b && b.left instanceof MulExpr) {
      return mul(b.left, mul(x,b.right));
    }
    else if (x instanceof MulExpr a) {
      return mul(a.left, mul(a.right,y));
    }
    else {
      return new MulExpr(x,y);
    }
  }

  static Expr powr( Expr x, Expr y ) {
    if (x instanceof ValExpr a && y instanceof ValExpr b) {
      return new ValExpr(pown(a.value,b.value));
    }
    else if (y instanceof ValExpr b && b.value == 0) {
      return new ValExpr(1);
    }
    else if (y instanceof ValExpr b && b.value == 1) {
      return x;
    }
    else if (x instanceof ValExpr a && a.value == 0) {
      return new ValExpr(0);
    }
    else {
      return new PowExpr(x,y);
    }
  }

  static Expr ln( Expr x ) {
    if (x instanceof ValExpr a && a.value == 1) {
      return new ValExpr(0);
    }
    else {
      return new LnExpr(x);
    }
  }

  static Expr d( String x, Expr e  ) {
    if (e instanceof ValExpr) {
      return new ValExpr(0);
    }
    else if (e instanceof VarExpr a) {
      return new ValExpr(a.name == x ? 1 : 0);
    }
    else if (e instanceof AddExpr a) {
      Expr f = a.left;
      Expr g = a.right;
      return add(d(x,f),d(x,g));
    }
    else if (e instanceof MulExpr a) {
      Expr f = a.left;
      Expr g = a.right;
      return add(mul(f,d(x,g)),mul(g,d(x,f)));
    }
    else if (e instanceof PowExpr a) {
      Expr f = a.left;
      Expr g = a.right;
      return mul(powr(f,g),add(mul(mul(g,d(x,f)),powr(f,new ValExpr(-1))),mul(ln(f),d(x,g))));
    }
    else if (e instanceof LnExpr a) {
      Expr f = a.expr;
      return mul(d(x,f),powr(f,new ValExpr(-1)));
    }
    else {
      return e;
    }
  }

  static long count( Expr e  ) {
    if (e instanceof ValExpr) {
      return 1;
    }
    else if (e instanceof VarExpr) {
      return 1;
    }
    else if (e instanceof AddExpr a) {
      Expr f = a.left;
      Expr g = a.right;
      return count(f) + count(g);
    }
    else if (e instanceof MulExpr a) {
      Expr f = a.left;
      Expr g = a.right;
      return count(f) + count(g);
    }
    else if (e instanceof PowExpr a) {
      Expr f = a.left;
      Expr g = a.right;
      return count(f) + count(g);
    }
    else if (e instanceof LnExpr a) {
      Expr f = a.expr;
      return count(f);
    }
    else {
      return 0;
    }
  }

  static Expr deriv( long i, Expr e) {
    Expr f = d("x",e);
    System.out.println( (i+1) + " count: " + count(f) );
    return f;
  }

  static Expr nest( long s, Expr e) {
    long n = s;
    while(n > 0) {
      e = deriv(s - n, e);
      n--;
    }
    return e;
  }

  public static void main(String args[])
  {
    Expr x = new VarExpr("x");
    Expr e = powr(x,x);
    nest(10,e);
    System.out.println( "done" );
  }
}

File: test/bench/cpp/deriv.cpp
#include <math.h>
#include <string.h>
#include <iostream>

enum Kind {
  Val,
  Var,
  Add,
  Mul,
  Pow,
  Ln
};

class Expr {
public:
  Kind kind;
  Expr(Kind k) {
    this->kind = k;
  }
};

class ValExpr : public Expr {
public:
  long value;
  ValExpr(long i) : Expr(Val) {
    this->value = i;
  }
};

class VarExpr : public Expr {
public:
  const char* name;
  VarExpr( const char* n ) : Expr(Var) {
    this->name = n;
  }
};

class UnaryExpr : public Expr {
public:
  const Expr* expr;
  UnaryExpr( Kind k, const Expr* e ) : Expr(k) {
    this->expr = e;
  }
};

class BinExpr : public Expr {
public:
  const Expr* left;
  const Expr* right;
  BinExpr( Kind k, const Expr* e1, const Expr* e2 ) : Expr(k) {
    this->left = e1;
    this->right = e2;
  }

};

static long pown(long x, long n) {
  if (n==0) return 1;
  else if (n == 1) return x;
  else {
    long y = pown(x, n/2);
    return (y * y * (n%2 == 0 ? 1 : x));
  }
}

static const Expr* add( const Expr* x, const Expr* y ) {
  if (x->kind == Val && y->kind == Val) {
    return new ValExpr(((ValExpr*)x)->value + ((ValExpr*)y)->value);
  }
  else if (x->kind==Val && ((ValExpr*)x)->value==0) {
    return y;
  }
  else if (y->kind==Val && ((ValExpr*)y)->value==0) {
    return x;
  }
  else if (y->kind==Val) {
    return add(y,x);
  }
  else if (x->kind==Val && y->kind==Add && ((BinExpr*)y)->left->kind==Val) {
    long lval = ((ValExpr*)((BinExpr*)y)->left)->value;
    return add(new ValExpr(((ValExpr*)x)->value + lval), ((BinExpr*)y)->right);
  }
  else if (y->kind==Add && ((BinExpr*)y)->left->kind==Val) {
    return add(((BinExpr*)y)->left,add(x,((BinExpr*)y)->right));
  }
  else if (x->kind==Add) {
    return add(((BinExpr*)x)->left,add(((BinExpr*)x)->right,y));
  }
  else {
    return new BinExpr(Add,x,y);
  }
}

static const Expr* mul( const Expr* x, const Expr* y ) {
  if (x->kind == Val && y->kind == Val) {
    return new ValExpr(((ValExpr*)x)->value * ((ValExpr*)y)->value);
  }
  else if (x->kind==Val && ((ValExpr*)x)->value==0) {
    return x;
  }
  else if (y->kind==Val && ((ValExpr*)y)->value==0) {
    return y;
  }
  else if (x->kind==Val && ((ValExpr*)x)->value==1) {
    return y;
  }
  else if (y->kind==Val && ((ValExpr*)y)->value==1) {
    return x;
  }
  else if (y->kind==Val) {
    return mul(y,x);
  }
  else if (x->kind==Val && y->kind==Mul && ((BinExpr*)y)->left->kind==Val) {
    long lval = ((ValExpr*)((BinExpr*)y)->left)->value;
    return mul(new ValExpr(((ValExpr*)x)->value * lval), ((BinExpr*)y)->right);
  }
  else if (y->kind==Mul && ((BinExpr*)y)->left->kind==Val) {
    return mul(((BinExpr*)y)->left,mul(x,((BinExpr*)y)->right));
  }
  else if (x->kind==Mul) {
    return mul(((BinExpr*)x)->left,mul(((BinExpr*)x)->right,y));
  }
  else {
    return new BinExpr(Mul,x,y);
  }
}

static const Expr* powr( const Expr* x, const Expr* y) {
  if (x->kind == Val && y->kind == Val) {
    return new ValExpr( pown(((ValExpr*)x)->value,((ValExpr*)y)->value));
  }
  else if (y->kind==Val && ((ValExpr*)y)->value == 0) {
    return new ValExpr(1);
  }
  else if (y->kind==Val && ((ValExpr*)y)->value == 1) {
    return x;
  }
  else if (x->kind==Val && ((ValExpr*)x)->value == 0) {
    return new ValExpr(0);
  }
  else {
    return new BinExpr(Pow,x,y);
  }
}

static const Expr* ln(const Expr* n) {
  if (n->kind == Val && ((ValExpr*)n)->value == 1) {
    return new ValExpr(0);
  }
  else {
    return new UnaryExpr(Ln,n);
  }
}

static const Expr* d( const char* x, const Expr* e) {
  if (e->kind == Val) {
    return new ValExpr(0);
  }
  else if (e->kind==Var) {
    return new ValExpr( strcmp(((VarExpr*)e)->name,x)==0 ? 1 : 0);
  }
  else if (e->kind==Add) {
    const Expr* f = ((BinExpr*)e)->left;
    const Expr* g = ((BinExpr*)e)->right;
    return add(d(x,f),d(x,g));
  }
  else if (e->kind==Mul) {
    const Expr* f = ((BinExpr*)e)->left;
    const Expr* g = ((BinExpr*)e)->right;
    return add(mul(f,d(x,g)),mul(g,d(x,f)));
  }
  else if (e->kind==Pow) {
    const Expr* f = ((BinExpr*)e)->left;
    const Expr* g = ((BinExpr*)e)->right;
    return  mul(powr(f,g),add(mul(mul(g,d(x,f)),powr(f,new ValExpr(-1))),mul(ln(f),d(x,g))));
  }
  else { // if (e->kind==Ln) {
    const Expr* f = ((UnaryExpr*)e)->expr;
    return mul(d(x,f),powr(f,new ValExpr(-1)));
  }
}

static long count( const Expr* e) {
  if (e->kind == Val) {
    return 1;
  }
  else if (e->kind==Var) {
    return 1;
  }
  else if (e->kind==Add) {
    const Expr* f = ((BinExpr*)e)->left;
    const Expr* g = ((BinExpr*)e)->right;
    return count(f) + count(g);
  }
  else if (e->kind==Mul) {
    const Expr* f = ((BinExpr*)e)->left;
    const Expr* g = ((BinExpr*)e)->right;
    return count(f) + count(g);
  }
  else if (e->kind==Pow) {
    const Expr* f = ((BinExpr*)e)->left;
    const Expr* g = ((BinExpr*)e)->right;
    return count(f) + count(g);
  }
  else { // if (e->kind==Ln) {
    const Expr* f = ((UnaryExpr*)e)->expr;
    return count(f);
  }
}

static const Expr* deriv(long i, const Expr* e) {
  const Expr* f = d("x",e);
  std::cout << (i+1) << " count: " << count(f) << "\n";
  return f;
}

static const Expr* nest( long s, const Expr* e) {
  long n = s;
  while(n > 0) {
    e = deriv(s - n, e);
    n--;
  }
  return e;
}


int main(int argc, char ** argv) {
  const Expr* x = new VarExpr("x");
  const Expr* e = powr(x,x);
  nest(10,e);
  std::cout << "done\n";
  return 0;
}
